File: app/src/main/java/by/toggi/rxbsuir/fragment/SettingsFragment.java
package by.toggi.rxbsuir.fragment;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import com.google.android.material.timepicker.MaterialTimePicker;

import java.time.LocalTime;

import javax.inject.Inject;
import javax.inject.Named;

import by.toggi.rxbsuir.BuildConfig;
import by.toggi.rxbsuir.PreferenceHelper;
import by.toggi.rxbsuir.R;
import by.toggi.rxbsuir.Utils;
import dagger.hilt.android.AndroidEntryPoint;
import rx.Subscription;

@AndroidEntryPoint
public class SettingsFragment extends PreferenceFragmentCompat
        implements Preference.OnPreferenceChangeListener, Preference.OnPreferenceClickListener {

    @Inject
    com.f2prateek.rx.preferences.Preference<LocalTime> mLocalTimePreference;
    @Inject
    @Named(PreferenceHelper.FAVORITE_SYNC_ID)
    com.f2prateek.rx.preferences.Preference<String> mFavoriteSyncIdPrerefence;

    private Subscription mSubscription;
    private Preference mNotificationTimePreference;

    @Override
    public void onCreatePreferences(@Nullable Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preferences, rootKey);

        findPreference("build_version").setSummary(BuildConfig.VERSION_NAME);
        findPreference("rate_app").setOnPreferenceClickListener(this);
        findPreference("privacy_policy").setOnPreferenceClickListener(this);
        findPreference(PreferenceHelper.IS_TODAY_ENABLED).setOnPreferenceChangeListener(this);
        findPreference(PreferenceHelper.ARE_CIRCLES_COLORED).setOnPreferenceChangeListener(this);
        findPreference(PreferenceHelper.IS_FAM_ENABLED).setOnPreferenceChangeListener(this);
        mNotificationTimePreference = findPreference("notification_time");
        mNotificationTimePreference.setEnabled(mFavoriteSyncIdPrerefence.get() != null);
        mNotificationTimePreference.setOnPreferenceClickListener(preference -> {
            var localTime = mLocalTimePreference.get();
            var picker = new MaterialTimePicker.Builder()
                    .setHour(localTime.getHour())
                    .setMinute(localTime.getMinute())
                    .build();
            picker.addOnPositiveButtonClickListener(v -> mLocalTimePreference.set(LocalTime.of(picker.getHour(), picker.getMinute())));
            picker.show(getChildFragmentManager(), null);
            return true;
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        mSubscription = mLocalTimePreference.asObservable().subscribe(localTime -> {
            if (mNotificationTimePreference != null) {
                mNotificationTimePreference.setSummary(localTime.toString());
            }
            if (mFavoriteSyncIdPrerefence.get() != null) {
                Utils.setNotificationAlarm(getActivity(), localTime);
            }
        });
    }

    @Override
    public void onPause() {
        super.onPause();
        Utils.unsubscribe(mSubscription);
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        switch (preference.getKey()) {
            case PreferenceHelper.IS_TODAY_ENABLED:
            case PreferenceHelper.ARE_CIRCLES_COLORED:
            case PreferenceHelper.IS_FAM_ENABLED:
                Utils.restartApp(getActivity());
                break;
        }
        return true;
    }

    @Override
    public boolean onPreferenceClick(Preference preference) {
        switch (preference.getKey()) {
            case "rate_app":
                Utils.openPlayStorePage(getActivity());
                return true;
            case "privacy_policy":
                Utils.openPrivacyPolicyPage(getActivity());
                return true;
        }
        return false;
    }
}

File: app/src/main/kotlin/by/toggi/rxbsuir/settings/SettingsFragment.kt
package by.toggi.rxbsuir.settings

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import by.toggi.rxbsuir.PreferenceHelper
import by.toggi.rxbsuir.R
import by.toggi.rxbsuir.Utils
import by.toggi.rxbsuir.dagger.PerFragment
import by.toggi.rxbsuir.util.consume
import com.takisoft.fix.support.v7.preference.PreferenceFragmentCompatDividers
import com.takisoft.fix.support.v7.preference.TimePickerPreference
import dagger.android.ContributesAndroidInjector
import dagger.android.support.AndroidSupportInjection
import hu.supercluster.paperwork.Paperwork
import org.threeten.bp.LocalTime
import rx.subscriptions.CompositeSubscription
import javax.inject.Inject
import javax.inject.Named
import com.f2prateek.rx.preferences.Preference as RxPreference

class SettingsFragment : PreferenceFragmentCompatDividers() {

  private val subscriptions = CompositeSubscription()

  @Inject
  lateinit var paperwork: Paperwork
  @Inject
  lateinit var localTimePreference: RxPreference<LocalTime>
  @field:[Inject Named(PreferenceHelper.FAVORITE_SYNC_ID)]
  lateinit var favoriteSyncIdPreference: RxPreference<String>

  private lateinit var notificationTimePreference: TimePickerPreference

  override fun onAttach(context: Context?) {
    AndroidSupportInjection.inject(this)
    super.onAttach(context)
  }

  override fun onCreatePreferencesFix(savedInstanceState: Bundle?, rootKey: String?) {
    setPreferencesFromResource(R.xml.preferences, rootKey)

    findPreference("build_version").summary = paperwork["gitInfo"]
    findPreference("rate_app").setOnPreferenceClickListener {
      consume { Utils.openPlayStorePage(context) }
    }

    notificationTimePreference = findPreference("notification_time") as? TimePickerPreference
        ?: throw IllegalStateException("notification_time == null.")
    notificationTimePreference.isEnabled = favoriteSyncIdPreference.get() != null
  }

  override fun onStart() {
    super.onStart()
    subscriptions.add(localTimePreference
        .asObservable()
        .subscribe {
          notificationTimePreference.setTime(it.hour, it.minute)
          if (favoriteSyncIdPreference.get() != null) {
            Utils.setNotificationAlarm(context, it)
          }
        })
  }

  override fun onStop() {
    subscriptions.clear()
    super.onStop()
  }

  companion object {

    fun newInstance(): Fragment {
      return SettingsFragment()
    }
  }

  @dagger.Module
  interface Module {

    @PerFragment
    @ContributesAndroidInjector
    fun contribute(): SettingsFragment
  }
}
